src/api/offline/ledgerhash.js

```diff
@@ -28,25 +28,47 @@ function hashLedgerHeader(ledgerHeader) {
   return common.core.Ledger.calculateLedgerHash(header);
 }
 
-function computeLedgerHash(ledgerHeader: Object, transactions: Array<Object>
-): string {
-  if (transactions) {
-    const txs = _.map(transactions, tx => {
-      const mergeTx = _.assign({}, _.omit(tx, 'tx'), tx.tx || {});
-      const renameMeta = _.assign({}, _.omit(mergeTx, 'meta'),
-        tx.meta ? {metaData: tx.meta} : {});
-      return renameMeta;
-    });
-    const ledger = common.core.Ledger.from_json({transactions: txs});
-    const transactionHash = ledger.calc_tx_hash().to_hex();
-    if (ledgerHeader.transaction_hash !== undefined
-        && ledgerHeader.transaction_hash !== transactionHash) {
-      throw new common.errors.ValidationError('transaction_hash in header'
-        + ' does not match computed hash of transactions');
-    }
-    return hashLedgerHeader(_.assign({}, ledgerHeader, {transactionHash}));
+function computeTransactionHash(ledger) {
+  if (ledger.rawTransactions === undefined) {
+    return ledger.transactionHash;
   }
-  return hashLedgerHeader(ledgerHeader);
+  const transactions = JSON.parse(ledger.rawTransactions);
+  const txs = _.map(transactions, tx => {
+    const mergeTx = _.assign({}, _.omit(tx, 'tx'), tx.tx || {});
+    const renameMeta = _.assign({}, _.omit(mergeTx, 'meta'),
+      tx.meta ? {metaData: tx.meta} : {});
+    return renameMeta;
+  });
+  const ledgerObject = common.core.Ledger.from_json({transactions: txs});
+  const transactionHash = ledgerObject.calc_tx_hash().to_hex();
+  if (ledger.transactionHash !== undefined
+      && ledger.transactionHash !== transactionHash) {
+    throw new common.errors.ValidationError('transactionHash in header'
+      + ' does not match computed hash of transactions');
+  }
+  return transactionHash;
+}
+
+function computeStateHash(ledger) {
+  if (ledger.rawState === undefined) {
+    return ledger.stateHash;
+  }
+  const state = JSON.parse(ledger.rawState);
+  const ledgerObject = common.core.Ledger.from_json({accountState: state});
+  const stateHash = ledgerObject.calc_account_hash().to_hex();
+  if (ledger.stateHash !== undefined && ledger.stateHash !== stateHash) {
+    throw new common.errors.ValidationError('stateHash in header'
+      + ' does not match computed hash of state');
+  }
+  return stateHash;
+}
+
+function computeLedgerHash(ledger: Object): string {
```

Comment:
What I'm trying to get at is that there's this new ripple api which which
has a request/response interface speaking json primitives for parameters
and return values that would allow it to be easily exported from an RPC
server.

There's exports like generateWallet, where the analogous rippled command
"wallet_create" will only serve to local connections. computeLedgerHash
seems somewhat in the same class. Both could feasibly be seen to be
exported from a rippled (or a satellite app if rippled became p2p txn
processing only) served json rpc API.

But there's little guys like isConnected, connect, disconnect which are
clearly `client` routines.

Take for example the getTransactions api, where it's conceivable people may
want an event handler to be executed for each page of transactions as they
come in to allow them to update a UI. They might build a wrapper around the
getTransactions service. That would be a `client` routine.

And `errors`?

I think if say in the src/api/index.js what are server and what are client
routines were clearly marked out, it would help communicate the ultimate
intent of some of these apis one day being accessed via RPC which unless I
got confused somewhere is the goal?

It's getting a little weird to see things like
RippleAPI.prototype.isValidAddress, (which would likely not be exported
from a json rpc) sort of tacked on as a method, seemingly only because
RippleAPI is currently the only namespace exported.